src/Functional.CQS.AOP.IoC.SimpleInjector.MetricsCapturing/ConventionBasedDecoratorRegistrationGatewayExtensions.cs

```diff
@@ -0,0 +1,98 @@
+				return;
+
+			var queryAndResultTypeWithMetricsCapturingStrategyDefinedCollection = new HashSet<QueryAndResultType>(gateway.AssemblyCollection
+				.SelectMany(assembly => assembly.GetTypes().Where(t => t.IsMetricsCapturingStrategyForQueryType()))
+				.Select(x => x.GetGenericParametersForQueryMetricsCapturingStrategyType())
+				.WhereSome());
+
+			bool hasMetricsCapturingStrategyDefinedForQuery(DecoratorPredicateContext c) => c.ToServiceAndImplementationType().HasMetricsCapturingStrategyDefined(queryAndResultTypeWithMetricsCapturingStrategyDefinedCollection);
+			gateway.Container.RegisterSingleton(typeof(IMetricsCapturingStrategyForQuery<,>), gateway.AssemblyCollection);
+			gateway.Container.RegisterDecorator(typeof(IQueryHandler<,>), typeof(QueryHandlerMetricsCapturingDecorator<,>), gateway.Lifestyle, hasMetricsCapturingStrategyDefinedForQuery);
+			gateway.Container.RegisterDecorator(typeof(IAsyncQueryHandler<,>), typeof(AsyncQueryHandlerMetricsCapturingDecorator<,>), gateway.Lifestyle, hasMetricsCapturingStrategyDefinedForQuery);
+		}
+
+		private static void RegisterMetricsCapturingDecoratorForIndividualCommandHandlerImplementations(this ConventionBasedDecoratorRegistrationGateway gateway, MetricsCapturingModuleConfigurationParameters configurationParameters)
+		{
+			if (!configurationParameters.CommandSpecificMetricsCapturingDecoratorEnabled)
+				return;
+
+			var commandAndErrorTypeWithCachingStrategyDefinedCollection = new HashSet<CommandAndErrorType>(gateway.AssemblyCollection
+				.SelectMany(assembly => assembly.GetTypes().Where(t => t.IsMetricsCapturingStrategyForCommandType()))
+				.Select(x => x.GetGenericParametersForCommandMetricsCapturingStrategyType())
+				.WhereSome());
+
+			bool hasMetricsCapturingStrategyDefinedForCommand(DecoratorPredicateContext c) => c.ToServiceAndImplementationType().HasMetricsCapturingStrategyDefined(commandAndErrorTypeWithCachingStrategyDefinedCollection);
+			gateway.Container.RegisterSingleton(typeof(IMetricsCapturingStrategyForCommand<,>), gateway.AssemblyCollection);
+			gateway.Container.RegisterDecorator(typeof(ICommandHandler<,>), typeof(CommandHandlerMetricsCapturingDecorator<,>), gateway.Lifestyle, hasMetricsCapturingStrategyDefinedForCommand);
+			gateway.Container.RegisterDecorator(typeof(IAsyncCommandHandler<,>), typeof(AsyncCommandHandlerMetricsCapturingDecorator<,>), gateway.Lifestyle, hasMetricsCapturingStrategyDefinedForCommand);
+		}
+
+		private static void RegisterUniversalMetricsCapturingDecorator<TUniversalMetricsCapturingStrategy>(this ConventionBasedDecoratorRegistrationGateway gateway, MetricsCapturingModuleConfigurationParameters configurationParameters)
+			where TUniversalMetricsCapturingStrategy : class, IUniversalMetricsCapturingStrategy
+		{
+			if (!configurationParameters.UniversalMetricsCapturingDecoratorEnabled)
+				return;
+
+			gateway.Container.RegisterSingleton<IUniversalMetricsCapturingStrategy, TUniversalMetricsCapturingStrategy>();
```

Comment:
For all 3 registration methods in this class, I assume that we always want the strategies to be registered as singleton, regardless of whether `gateway.Lifestyle` is singleton or transient?

Reply:
Ah, looking at your documentation, this is intentional. 👍